src/server/pkg/sync/sync.go

```diff
@@ -179,20 +196,56 @@ func (p *Puller) Pull(client *pachclient.APIClient, root string, repo, commit, f
 			})
 		}
 		if emptyFiles {
-			return p.makeFile(path, func(w io.Writer) error { return nil })
+			_, err := p.makeFile(path, false, func(w io.Writer) error { return nil })
+			return err
 		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	limiter := limit.New(concurrency)
+	var eg errgroup.Group
+	if !pipes && !emptyFiles {
 		eg.Go(func() (retErr error) {
 			limiter.Acquire()
 			defer limiter.Release()
-			return p.makeFile(path, func(w io.Writer) error {
-				return client.GetFile(repo, commit, fileInfo.File.Path, 0, 0, w)
-			})
+			var oldFile *os.File
+			// for the common case, when neither pipes nor emptyFiles are set, we make a recursive
+			// GetFiles call in order to retrieve all the data for the repo from object storage in a single call.
+			if err := client.GetFiles(repo, commit, file, 0, 0, func(fileInfo *pfs.FileInfo, r io.Reader) error {
+				if fileInfo != nil && fileInfo.File != nil {
```

Comment:
This whole file creation/closing mechanism feels a little janky. Right now we create a file for every new file we get streamed back from the `GetFiles` call and then just leave it open so that we can continue copying data to it if the following messages from `GetFiles` are continuations of the file.

This means we become responsible for closing the file outside of the `makeFile` call. This seems like a working solution but lemme know if there's a better way.

Reply:
That seems fine.